src/identity/processor.rs

```diff
@@ -529,19 +529,31 @@ impl IdentityProcessor for OffChainIdentityProcessor {
         processed_tree: &TreeVersion<Intermediate>,
         mined_tree: &TreeVersion<Canonical>,
     ) -> anyhow::Result<()> {
-        let batches = {
-            let mut committed_batches = self.committed_batches.lock().unwrap();
-            let copied = committed_batches.clone();
-            committed_batches.clear();
-            copied
-        };
-
-        for batch in batches.iter() {
+        loop {
+            let batch = {
+                let mut committed_batches = self.committed_batches.lock().unwrap();
+                committed_batches.pop_front()
+            };
+
+            let Some(batch) = batch else {
+                return Ok(());
+            };
+
             let mut tx = self
                 .database
                 .begin_tx(IsolationLevel::ReadCommitted)
                 .await?;
 
+            let root_state = tx.get_root_state(&batch.next_root).await?;
+            if root_state.is_none() {
+                // If root is not in identities table we can't mark it as processed or mined.
+                // It happens sometimes as we do not have atomic operation for database and tree
+                // insertion.
+                // TODO: check if this is still possible after HA being done
+                tx.commit().await?;
```

Comment:
wouldn't rollback be more fitting here? Since we intend for nothing to happen?

Reply:
I would rather leave commit here as there was no error at all.